File: draft_design_pattern/bridge.py
# bridgeパターンは、継承ではなくcompositionを優先するためのものです。
# ある階層の実装の詳細は、別の階層にある別のオブジェクトにプッシュされます
from abc import ABCMeta, abstractmethod

# WebPageという階層を設定します


class WebPage(metaclass=ABCMeta):
    def __init__(self, theme: "Theme") -> None:
        self.theme = theme

    @abstractmethod
    def getContent(self) -> str:
        pass


class About(WebPage):
    def __init__(self, theme: "Theme") -> None:
        super().__init__(theme)

    def getContent(self) -> str:
        return self.theme.getColor() + "のAboutページ"


class Careers(WebPage):
    def __init__(self, theme: "Theme") -> None:
        super().__init__(theme)

    def getContent(self) -> str:
        return self.theme.getColor() + "のCareersページ"


# Themeという階層を記述します

class Theme(metaclass=ABCMeta):
    @abstractmethod
    def getColor(self) -> str:
        pass


class DarkTheme(Theme):
    def getColor(self) -> str:
        return 'Dark Black'


class LightTheme(Theme):
    def getColor(self) -> str:
        return 'Off white'


class AquaTheme(Theme):
    def getColor(self) -> str:
        return 'Light blue'


# 2つの階層を次のように使えます

darkTheme = DarkTheme()

about = About(darkTheme)
careers = Careers(darkTheme)

print(about.getContent())
print(careers.getContent())

File: draft_design_pattern/bridge.ts
export {};

interface Theme {
  getColor(): string;
}

class DarkTheme implements Theme {
  getColor(): string {
    return "Dark Black";
  }
}
class LightTheme implements Theme {
  getColor(): string {
    return "Off white";
  }
}
class AquaTheme implements Theme {
  getColor(): string {
    return "Light blue";
  }
}

interface WebPage {
  getContent(): string;
}

class About implements WebPage {
  constructor(private theme: Theme) {}
  getContent(): string {
    return this.theme.getColor() + "のAboutページ";
  }
}

class Careers implements WebPage {
  constructor(private theme: Theme) {}
  getContent(): string {
    return this.theme.getColor() + "のCareersページ";
  }
}

const darkTheme = new DarkTheme();
let about = new About(darkTheme);
let careers = new Careers(darkTheme);

console.log(about.getContent());
console.log(careers.getContent());

const lightTheme = new LightTheme();
about = new About(lightTheme);
careers = new Careers(lightTheme);

console.log(about.getContent());
console.log(careers.getContent());

const aquaTheme = new AquaTheme();
about = new About(aquaTheme);
careers = new Careers(aquaTheme);

console.log(about.getContent());
console.log(careers.getContent());
